akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventsBySlicesFirehose.scala

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (C) 2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+package akka.projection.grpc.internal
+
+import java.time.Instant
+import java.time.{ Duration => JDuration }
+import java.util.UUID
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicLong
+
+import scala.annotation.nowarn
+
+import akka.NotUsed
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.Extension
+import akka.actor.typed.ExtensionId
+import akka.persistence.query.Offset
+import akka.persistence.query.PersistenceQuery
+import akka.persistence.query.TimestampOffset
+import akka.persistence.query.typed.EventEnvelope
+import akka.persistence.query.typed.scaladsl.EventsBySliceQuery
+import akka.stream.KillSwitch
+import akka.stream.KillSwitches
+import akka.stream.scaladsl.BroadcastHub
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.Source
+import org.slf4j.LoggerFactory
+
+object EventsBySlicesFirehose extends ExtensionId[EventsBySlicesFirehose] {
+  private val log = LoggerFactory.getLogger(classOf[EventsBySlicesFirehose])
+
+  def createExtension(system: ActorSystem[_]): EventsBySlicesFirehose =
+    new EventsBySlicesFirehose(system)
+
+  // Java API
+  def get(system: ActorSystem[_]): EventsBySlicesFirehose = apply(system)
+
+  final case class FirehoseKey(entityType: String, sliceRange: Range)
+
+  final case class ConsumerTracking(count: Long, timestamp: Instant)
+
+  class Firehose(val firehoseHub: Source[EventEnvelope[Any], NotUsed]) {
+    val inCount: AtomicLong = new AtomicLong
+    val consumerTracking: ConcurrentHashMap[String, ConsumerTracking] = new ConcurrentHashMap
+
+    def updateConsumerTracking(consumerId: String, timestamp: Instant): Boolean = {
+      val tracking = consumerTracking.get(consumerId) match {
+        case null =>
+          ConsumerTracking(count = 1, timestamp)
+        case existing =>
+          if (timestamp.compareTo(existing.timestamp) > 0)
+            ConsumerTracking(count = existing.count + 1, timestamp)
+          else
+            existing.copy(count = existing.count + 1)
+      }
+      // FIXME trace
+      log.debug("Consumer [{}] count [{}], in count [{}]", consumerId, tracking.count, inCount.get)
+      // no concurrent updates for a given consumerId
+      consumerTracking.put(consumerId, tracking)
+      true // FIXME detect slow consumer
```

Comment:
The idea is that when the gap between inCount and the consumer count is near the BroadCastHub buffer size there might be some slow consumers. Maybe act after a while in case it's just temporary. Then it can look at the counts from all consumers in the `consumerTracking` Map to find the slowest.

Included the timestamp there, but that might not be needed for the decision.